Shield CLI Sends Null Request Body if User Prompted For Auth

POST requests that require request bodys will fail if auth is requested of the user.
First thing that comes to mind is that the request body is being read on the first request attempt, and as the second attempt is made, the cursor of the body reader has reached the end.

```
→  shield create target -DT
DEBUG> shield cli starting up
DEBUG> enabling TRACE output
DEBUG> Command: 'create target'
Using http://localhost:8181 (test) as SHIELD backend

DEBUG> running 'create target' command
Target Name: beep
Summary: boop
Plugin Name: postgres
Configuration: {"endpoint","schmendpoint"}
Remote IP:port: 127.0.0.1:5444


Target Name:    beep
Summary:        boop
Plugin Name:    postgres
Configuration:  {"endpoint","schmendpoint"}
Remote IP:port: 127.0.0.1:5444


Really create this target? [y/n] y
DEBUG> JSON:
DEBUG>   {"agent":"127.0.0.1:5444","endpoint":"{\"endpoint\",\"schmendpoint\"}","name":"beep","plugin":"postgres","summary":"boop"}
DEBUG>
Request:
POST /v1/targets HTTP/1.1
Host: localhost:8181
Content-Type: application/json

{"agent":"127.0.0.1:5444","endpoint":"{\"endpoint\",\"schmendpoint\"}","name":"beep","plugin":"postgres","summary":"boop"}
---------------------------
Response:
HTTP/1.1 401 Unauthorized
Content-Length: 12
Content-Type: text/plain; charset=utf-8
Date: Mon, 13 Jun 2016 20:22:42 GMT
Www-Authenticate: Basic realm="shield"

Unauthorized
---------------------------
Authentication Required

User: testuser

Password:

Request:
POST /v1/targets HTTP/1.1
Host: localhost:8181
Authorization: Basic dGVzdHVzZXI6dGVzdHBhc3M=
Content-Type: application/json


---------------------------
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x2b851e]

goroutine 1 [running]:
panic(0x423940, 0xc8200120f0)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
net/http/httputil.DumpResponse(0x0, 0xc800000001, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/httputil/dump.go:285 +0x42e
github.com/starkandwayne/shield/api.debugResponse(0x0)
	/Users/tom/go_workspace/src/github.com/starkandwayne/shield/api/url.go:82 +0x47
github.com/starkandwayne/shield/api.makeRequest(0xc820164000, 0xc8201441e0, 0x0, 0x0)
	/Users/tom/go_workspace/src/github.com/starkandwayne/shield/api/url.go:174 +0x368
github.com/starkandwayne/shield/api.promptAndAuth(0xc820184000, 0xc820164000, 0xc8201380a0, 0x0, 0x0)
	/Users/tom/go_workspace/src/github.com/starkandwayne/shield/api/url.go:217 +0x5ca
github.com/starkandwayne/shield/api.(*URL).Request(0xc820162000, 0x36a240, 0xc820114690, 0xc820164000, 0x0, 0x0)
	/Users/tom/go_workspace/src/github.com/starkandwayne/shield/api/url.go:95 +0x100
github.com/starkandwayne/shield/api.(*URL).Post(0xc820162000, 0x36a240, 0xc820114690, 0xc82001c780, 0x7a, 0x0, 0x0)
	/Users/tom/go_workspace/src/github.com/starkandwayne/shield/api/url.go:144 +0x225
github.com/starkandwayne/shield/api.CreateTarget(0xc82001c780, 0x7a, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/tom/go_workspace/src/github.com/starkandwayne/shield/api/targets.go:41 +0xe9
main.main.func12(0xc8200131e0, 0xc820013208, 0xc820013228, 0xc820013258, 0xc820013278, 0xc820013298, 0xc8200132b8, 0xc8200132d8, 0xc8200132f8, 0xc820013318, ...)
	/Users/tom/go_workspace/src/github.com/starkandwayne/shield/cmd/shield/main.go:419 +0x34e
main.(*Command).do(0xc8200981c0, 0xc8200c8a20, 0x2, 0x2, 0xc820092400, 0x0, 0x0)
	/Users/tom/go_workspace/src/github.com/starkandwayne/shield/cmd/shield/command.go:141 +0x2db
main.(*Command).Execute(0xc8200981c0, 0xc8200c8a20, 0x2, 0x2, 0x0, 0x0)
	/Users/tom/go_workspace/src/github.com/starkandwayne/shield/cmd/shield/command.go:170 +0x50
main.main()
	/Users/tom/go_workspace/src/github.com/starkandwayne/shield/cmd/shield/main.go:1917 +0x4bf2
```

--- a/api/url.go
+++ b/api/url.go
@@ -5,13 +5,14 @@ import (
 	"crypto/tls"
 	"encoding/json"
 	"fmt"
-	"golang.org/x/crypto/ssh/terminal"
 	"io/ioutil"
 	"net/http"
 	"net/http/httputil"
 	"net/url"
 	"os"
 	"strings"
+
+	"golang.org/x/crypto/ssh/terminal"
 )
 
 type URL struct {
@@ -85,6 +86,19 @@ func debugResponse(res *http.Response) {
 }
 
 func (u *URL) Request(out interface{}, req *http.Request) error {
+	var bodyBytes []byte
+	var err error
+	if req.Body != nil {
+		bodyBytes, err = ioutil.ReadAll(req.Body)
+		if err != nil {
+			return err
+		}
+		req.Body = ioutil.NopCloser(bytes.NewReader(bodyBytes))
+		if err != nil {
+			return err
+		}
+	}
+
 	r, err := makeRequest(req)
 	if err != nil {
 		return err
@@ -92,6 +106,9 @@ func (u *URL) Request(out interface{}, req *http.Request) error {
 	defer r.Body.Close()
 
 	if r.StatusCode == 401 {
+		if req.Body != nil {
+			req.Body = ioutil.NopCloser(bytes.NewReader(bodyBytes))
+		}
 		r, err = promptAndAuth(r, req)
 		if err != nil {
 			return err

